### PR TITLE
ABI-Safety tool chain for react-native-win32.dll

### PR DESCRIPTION
**Pull request by aeulitz**
This change represents a draft of the integration of the C++/WinRT tool chain (xlang implementation for Windows/C++, as I understand it) into the build of react-native-win32.dll. With it, I'm aiming to get early feedback about the ABI integration and its build and test process.

The change does *not* provide a complete ABI for all the exports (only for one exemplary method, InitializeTracing). Rather, I'm aiming to create an example that area owners/experts can follow, helping to gradually flesh out the ABI. The change might remain in the respective branch until performance testing (binary size, number of symbols, call time overhead ...) shows that C++/WinRT is an acceptable ABI solution.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/2656)

### CODE REVIEW

File: vnext/Desktop.ABITests/ActivationFactory.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
+
```
Comment (aeulitz) [inline vnext/Desktop.ABITests/ActivationFactory.cpp]: REVIEW INFO: This file is boilerplate code provided by the C++/WinRT team. It represents a part of the C++/WinRT "runtime" that is currently being worked on, in particular a feature they call "registry-less activation". A variation of this code would likely need to be used in ReactNativeHost.

File: vnext/Desktop/React.Windows.Desktop.vcxproj [hunk 2]
```diff
@@ -65,11 +68,13 @@
         BOOST_ASIO_HAS_IOCP - Force unique layout/size for boost::asio::basic_stream_socket<> subtypes.
       -->
       <PreprocessorDefinitions>BOOST_ASIO_HAS_IOCP;_WINSOCK_DEPRECATED_NO_WARNINGS;_WIN32_WINNT=_WIN32_WINNT_WIN7;WIN32;_WINDOWS;REACTNATIVEWIN32_EXPORTS;FOLLY_NO_CONFIG;WIN32_LEAN_AND_MEAN;NOMINMAX;GLOG_NO_ABBREVIATED_SEVERITIES;_HAS_AUTO_PTR_ETC;CHAKRACORE;RN_PLATFORM=win32;RN_EXPORT=;JSI_EXPORT=;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <PreprocessorDefinitions>_SILENCE_ALL_CXX17_DEPRECATION_WARNINGS;%(PreprocessorDefinitions)</PreprocessorDefinitions>
       <DebugInformationFormat>ProgramDatabase</DebugInformationFormat>
       <AdditionalOptions>%(AdditionalOptions) /Zc:strictStrings /bigobj</AdditionalOptions>
       <AssemblerListingLocation>..\build\$(IntDir)\%(RelativeDir)</AssemblerListingLocation>
       <PrecompiledHeader>Use</PrecompiledHeader>
       <PrecompiledHeaderFile>pch.h</PrecompiledHeaderFile>
+      <LanguageStandard>stdcpp17</LanguageStandard>
```
Comment (aeulitz) [inline vnext/Desktop/React.Windows.Desktop.vcxproj]: I'm aware that @JunielKatarn's latest change should make this unnecessary. It should get removed in the next change update.

**Comment by JunielKatarn** [comment]
As discussed offline, consider using namespace `Microsoft::React` instead of `facebook::react`, as this is no intended to be a `react-native` subproject.

File: vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Debug|x64">
+      <Configuration>Debug</Configuration>
+      <Platform>x64</Platform>
+    </ProjectConfiguration>
+    <ProjectConfiguration Include="Release|x64">
+      <Configuration>Release</Configuration>
+      <Platform>x64</Platform>
+    </ProjectConfiguration>
+    <!-- Kept for Visual Studio compatibility. -->
```
Comment (JunielKatarn) [inline vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj]: Reminder: Please remove this comment.

File: vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj [hunk 4]
```diff
@@ -0,0 +1,150 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\packages\Microsoft.Windows.CppWinRT.2.0.190506.1\build\native\Microsoft.Windows.CppWinRT.props" Condition="Exists('..\packages\Microsoft.Windows.CppWinRT.2.0.190506.1\build\native\Microsoft.Windows.CppWinRT.props')" />
+  <ItemGroup Label="ProjectConfigurations">
+    <ProjectConfiguration Include="Debug|x86">
```
Comment (JunielKatarn) [inline vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj]: Reminder: remove "...|x86" ProjectConfiguration entries.

File: vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  </ItemDefinitionGroup>
+  <ItemDefinitionGroup Condition="'$(OSS_RN)'=='true'">
+    <ClCompile>
+      <PreprocessorDefinitions>OSS_RN;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+    </ClCompile>
+  </ItemDefinitionGroup>
+  <ItemDefinitionGroup Condition="'$(Configuration)'=='Debug'">
+    <ClCompile>
+      <PreprocessorDefinitions>_DEBUG;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <AdditionalOptions Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">/JMC- %(AdditionalOptions)</AdditionalOptions>
```
Comment (JunielKatarn) [inline vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj]: Remove.

File: vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  <Target Name="SetCppWinRTReferences">
+    <ItemGroup>
+      <!--
+      To increase resilience against build configuration changes, we could try to obtain the .winmd
+      path below from the Desktop.DLL project reference. The respective task should look similar to
+      'GetTargetFileName' MSBuild task in the 'GetRNDllPath' target below (and potentially share
+      build logic with it).
+      -->
+      <CppWinRTDynamicWinMDReferences Include="$(OutputPath)..\React.Windows.Desktop\facebook.react.winmd">
+        <WinMDPath>$(OutputPath)..\React.Windows.Desktop\facebook.react.winmd</WinMDPath>
```
Comment (JunielKatarn) [inline vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj]: [nit] Rename as Microsoft.React.winmd?

File: vnext/Desktop.IntegrationTests/React.Windows.Desktop.IntegrationTests.vcxproj [hunk 7]
```diff
@@ -61,6 +61,7 @@
       <AdditionalIncludeDirectories>$(VCInstallDir)UnitTest\include;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
       <DebugInformationFormat>ProgramDatabase</DebugInformationFormat>
       <UseFullPaths>true</UseFullPaths>
+      <LanguageStandard>stdcpp17</LanguageStandard>
```
Comment (JunielKatarn) [inline vnext/Desktop.IntegrationTests/React.Windows.Desktop.IntegrationTests.vcxproj]: [nit] Not needed.

File: vnext/Desktop.ABITests/ActivationFactory.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+{
+	*factory = nullptr;
+	std::wstring_view name{ WINRT_WindowsGetStringRawBuffer(classId, nullptr) };
+	HMODULE library{ nullptr };
+
+	if (starts_with(name, L"facebook.react."))
+	{
+		library = LoadLibraryExW(RNDLLPATH, NULL, LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR | LOAD_LIBRARY_SEARCH_DEFAULT_DIRS);
+		auto le = GetLastError();
+		le = le;
```
Comment (Khalef1) [inline vnext/Desktop.ABITests/ActivationFactory.cpp]: Needed?

File: vnext/Desktop/NativeTraceEventSource.cpp [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+{
+	namespace
+	{
+		::winrt::facebook::react::INativeTraceHandler g_abiHandler;
+		std::atomic<uint32_t> g_abiHandlerRegistrationCookie = 0;
+
+		class InternalHandler : public ::facebook::react::INativeTraceHandler
+		{
+		public:
+			virtual void JSBeginSection(const char* profileName, const char* args) noexcept override
```
Comment (Khalef1) [inline vnext/Desktop/NativeTraceEventSource.cpp]: add null checks for the parameters?

File: vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj (hunk 5 above)
Comment (aeulitz) [reply vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj]: Removed.

---
In reply to: [296964936](https://github.com/microsoft/react-native-windows/pull/2656#discussion_r296964936) [](ancestors = 296964936)

File: vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj (hunk 3 above)
Comment (aeulitz) [reply vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj]: Removed.

---
In reply to: [296964730](https://github.com/microsoft/react-native-windows/pull/2656#discussion_r296964730) [](ancestors = 296964730)

File: vnext/Desktop.IntegrationTests/React.Windows.Desktop.IntegrationTests.vcxproj (hunk 7 above)
Comment (aeulitz) [reply vnext/Desktop.IntegrationTests/React.Windows.Desktop.IntegrationTests.vcxproj]: Removed.

---
In reply to: [296965215](https://github.com/microsoft/react-native-windows/pull/2656#discussion_r296965215) [](ancestors = 296965215)

File: vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj (hunk 6 above)
Comment (aeulitz) [reply vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj]: Let's do this as a separate change.

---
In reply to: [296965025](https://github.com/microsoft/react-native-windows/pull/2656#discussion_r296965025) [](ancestors = 296965025)

File: vnext/Desktop.ABITests/ActivationFactory.cpp (hunk 8 above)
Comment (aeulitz) [reply vnext/Desktop.ABITests/ActivationFactory.cpp]: Removed.

---
In reply to: [298782488](https://github.com/microsoft/react-native-windows/pull/2656#discussion_r298782488) [](ancestors = 298782488)

File: vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj (hunk 4 above)
Comment (aeulitz) [reply vnext/Desktop.ABITests/React.Windows.Desktop.ABITests.vcxproj]: Removed.

---
In reply to: [296964811](https://github.com/microsoft/react-native-windows/pull/2656#discussion_r296964811) [](ancestors = 296964811)

**Comment by aeulitz** [comment]
Let's do this as a separate change.

---
In reply to: [505203335](https://github.com/microsoft/react-native-windows/pull/2656#issuecomment-505203335) [](ancestors = 505203335)

**Pull request merged**